### DISCUSSION TITLE
"Failed to load response data" on Browser when calling FastAPI via Fetch

### DISCUSSION THREAD

**Discussion by chand1012** [fastapi/fastapi#6608] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.responses import JSONResponse

# fast api init
app = FastAPI()

# make an echo function for testing
@app.get("/echo")
def echo(start_lat: float, start_lng: float, end_lat: float, end_lng: float, height: int = 300, width: int = 300):
    # return the parameters
    return JSONResponse({'start_lat': start_lat, 'start_lng': start_lng, 'end_lat': end_lat, 'end_lng': end_lng, 'height': height, 'width': width})
```

### Description

When I call the above endpoint via this fetch request, my browser says it returned a 200, but cannot display the data.
```javascript
const response = await fetch(
        `${API_URL}/echo?start_lat=${startLat}&start_lng=${startLng}&end_lat=${endLat}&end_lng=${endLng}&width=${options.width}&height=${options.height}`,
        {
            method: "GET",
            // disable cors
            mode: "no-cors",
        }
    );
```

This is what my browser says, as well as the headers.
![image](https://user-images.githubusercontent.com/3521582/136137951-db970379-5b7d-460a-a670-452a37ac1eb1.png)

![image](https://user-images.githubusercontent.com/3521582/136137977-c8f1a79b-817d-45de-8e4f-5859ec110f16.png)

I am using a version of the Chromium browser. When I call the above endpoint via a command line tool or even just go to the URL in my browser, the output looks fine.

```
$ http GET 'http://127.0.0.1:5000/echo?start_lat=40.208283036057104&start_lng=-82.88578412158346&end_lat=40.11303888887799&end_lng=-82.88166424853658&width=300&height=300'
HTTP/1.1 200 OK
content-length: 145
content-type: application/json
date: Wed, 06 Oct 2021 03:55:27 GMT
server: uvicorn

{
    "end_lat": 40.11303888887799,
    "end_lng": -82.88166424853658,
    "height": 300,
    "start_lat": 40.208283036057104,
    "start_lng": -82.88578412158346,
    "width": 300
}
```

The function given is just a generic example, but this happens with all endpoints when called via Fetch on my browser. This isn't isolated to Chromium either, as I had the same issue on Firefox on both Linux and Windows 10.

### Operating System

Linux

### FastAPI Version

0.68.1

### Python Version

Python 3.9.7

**Comment by ghandic** [comment]
You need to use CORSMiddleware

**Comment by oligond** [comment]
You already pointed out yourself that the issue seems to be related to the specific client environment you're sending your request from (Javascript running in a browser). So it's obviously not a FastAPI issue by nature.
That said, without further investigation, I'd see two possible debugging paths to follow:

1. Your problem might indeed be related to CORS (as @ghandic assumes, although the fact that you get a 200 response to me doesn't seem to support this assumption) - then you might want to dive deeper into CORS. As a good resource to get started, I would recommend looking into https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors
You might also evaluate what the FastAPI docs are able to offer to your investigation and wrt solutions to try: https://fastapi.tiangolo.com/tutorial/cors/
2. Alternatively, it's worth checking with another browser to see whether the problem is specific to the browser with which you observed the issue at first try. A very superficial web search gave me the impression that Chrome/Chromium might specifically produce this result. A frontend pro might be familiar with this...

Happy Debugging!

**Comment by chand1012** [comment, marked as answer]
It was definitely CORS related. I don't know why my browser refused to display the data unless I enabled CORS on FastAPI, and why FastAPI returned a 200, but enabling CORS fixed the issue.

```python
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse

# fast api init
app = FastAPI()

origins = ['*']
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# make an echo function for testing
@app.get("/echo")
def echo(start_lat: float, start_lng: float, end_lat: float, end_lng: float, height: int = 300, width: int = 300):
    # return the parameters
    return JSONResponse({'start_lat': start_lat, 'start_lng': start_lng, 'end_lat': end_lat, 'end_lng': end_lng, 'height': height, 'width': width})
```

Thanks for the help @oligond @ghandic !
